fingertip/util/filesystem.py

```diff
@@ -0,0 +1,99 @@
+# Licensed under GNU General Public License v3 or later, see COPYING.
+# Copyright (c) 2020 Red Hat, Inc., see CONTRIBUTORS.
+
+"""
+Helper functions for fingertip: Setup and cleanup filesystem
+"""
+
+import os
+import sys
+import shutil
+import subprocess
+
+from fingertip.util import log
+from fingertip.util.path import MACHINES, CACHE
+from fingertip.util.reflink import is_supported
+
+
+SETUP = os.getenv('FINGERTIP_SETUP', 'suggest')
+SIZE = os.getenv('FINGERTIP_SETUP_SIZE', '25G')
+
+
+def create_supported_fs(backing_file, size):
+    subprocess.run(['fallocate', '-l', size, backing_file], check=True)
+    subprocess.run(['mkfs.xfs', '-m', 'reflink=1', backing_file],
+                   check=True)
+
+
+def mount_supported_fs(backing_file, tgt):
+    # use warning to avoid rewriting the prompt with sudo prompt
+    log.warning('mounting a reflink-supported filesystem for image storage...')
+    tgt_uid, tgt_gid = os.stat(tgt).st_uid, os.stat(tgt).st_gid
+    subprocess.run(['sudo', 'mount', '-o', 'loop', backing_file, tgt],
+                   check=True)
+    mount_uid, mount_gid = os.stat(tgt).st_uid, os.stat(tgt).st_gid
+    if (tgt_uid, tgt_gid) != (mount_uid, mount_gid):
+        log.debug(f'fixing owner:group ({tgt_uid}:{tgt_gid})')
+        subprocess.run(['sudo', 'chown', f'{tgt_uid}:{tgt_gid}', tgt],
+                       check=True)
+
+def storage_schedule_cleanup():
+    # Do this only if fingertip is in PATH
+    if not shutil.which("fingertip"):
+        log.warning('No `fingertip` found in PATH. Not scheduling '
+                    'automatic cleanup')
+        return
+
+    subprocess.run(['systemd-run', '--user', '--on-active', '12h',
```

Comment:
@sahanaprasad07, he's stepping on your toes.

I think `--on-calendar` would be a better fit here, or a sizeable portion of the images will always be left on-disk. Pairing it with a `--unit` will allow to install just one timer no matter how you call it.

I'm not sure how I feel about the requirement of having it in the $PATH that systemd-run can pick up (on my laptop I use it through Nix environments, rey workers just use checkouts, @thrix probably swears by virtualenvs, etc), but, well, satisfying all these scenarios would be fragile or outright impossible.

Reply:
clearly I'm too slow :D  Also, I haven't done too much python.

@t184256 even with --on-calendar option, you will run it after 12h right (or every 12h)? similar to --on-active

See if you find this information useful -
To create a .timer and .service file, with options that can make it persistent

[sprasad@linux fingertip]$ cat fingtertip-cleanup.timer
[Unit]
Description=Fingertip machines cleanup

[Timer]
OnCalendar=daily
AccuracySec=12h
Persistent=true

[Install]
WantedBy=timers.target

[sprasad@linux fingertip]$ cat fingertip-cleanup.service
[Unit]
Description=fingertip-cleanup

[Service]
ExecStart=/bin/bash ~/.cache/fingertip/cleanup

Reply:
Thank you for input. This indeed sounds like a good idea to keep it a real unit, otherwise there is really no simple way to get the status of this ad-hoc timer. But I guess I will keep this for Monday.

Reply:
depending on what you need `--unit` might be real enough

Reply:
Right. I will certainly add `--unit`.

The requirement for having `fingertip` in the path is to be able to invoke it with the above command. If it is started from local git checkout, I would have to invoke it differently, which might or might work, possibly running broken `fingertip` development checkout (at "random" times), crashing, not finding the repository at all or doing even worse things. If others will come with way how to do this reliably from `poetry` or from local git checkouts, this can change.

Can you clarify why you believe the `--on-calendar` would work better? I think the difference from `--on-active` is only the source point in time from where the counting starts (in fixed time in day or in fixed time from timer load).

Reply:
I believe `--on-active` is one-off.

Reply:
`--on-active` is not one-off -- it is repetitive. `--on-startup` and `--on-boot` would be one-off, but `--on-active` repeats according to my searching (even if it is not explicitly stated in the man pages):
https://bbs.archlinux.org/viewtopic.php?pid=1746222#p1746222
(do not stop reading after first few answers -- the fastest is not always right)

Reply:
Could you please show how to make it recurrent? I've tried, e.g. `systemd-run --user --unit=test --on-active=2m /bin/true` and the timer disappears from `systemctl --user list-timers`. Also, I don't see much difference between the two if both are recurrent, maybe `--on-active` is a bit better due to the lack of rounding even without extra timer settings.

Reply:
Right. I modified it to `--on-calendar` and added few more sanity checks to avoid most common issues we could face.